File: 4h-unfolder-win/src/FourHUnfolder.Geometry/Algorithms/UnionFind.cs
namespace FourHUnfolder.Geometry.Algorithms;

/// <summary>
/// Path-compressed, union-by-rank Union-Find over a contiguous range [0, size).
/// </summary>
internal sealed class UnionFind
{
    private readonly int[] _parent;
    private readonly int[] _rank;

    public UnionFind(int size)
    {
        _parent = new int[size];
        _rank   = new int[size];
        for (int i = 0; i < size; i++) _parent[i] = i;
    }

    public int Find(int x)
    {
        if (_parent[x] != x) _parent[x] = Find(_parent[x]);
        return _parent[x];
    }

    /// Returns false if a and b were already in the same component.
    public bool Union(int a, int b)
    {
        a = Find(a); b = Find(b);
        if (a == b) return false;
        if (_rank[a] < _rank[b]) (a, b) = (b, a);
        _parent[b] = a;
        if (_rank[a] == _rank[b]) _rank[a]++;
        return true;
    }
}

File: 4h-unfolder-mac-swift/Sources/FourHUnfolderCore/Core/Graph/UnionFind.swift
// MARK: - Union-Find with path compression + union by rank

final class UnionFind {
    private var parent: [Int]
    private var rank: [Int]

    init(count: Int) {
        parent = Array(0..<count)
        rank   = Array(repeating: 0, count: count)
    }

    func find(_ x: Int) -> Int {
        if parent[x] != x { parent[x] = find(parent[x]) }   // path compression
        return parent[x]
    }

    @discardableResult
    func union(_ a: Int, _ b: Int) -> Bool {
        let ra = find(a), rb = find(b)
        guard ra != rb else { return false }
        if rank[ra] < rank[rb]      { parent[ra] = rb }
        else if rank[ra] > rank[rb] { parent[rb] = ra }
        else                        { parent[rb] = ra; rank[ra] += 1 }
        return true
    }
}
